wallet/wallet.go

```diff
@@ -313,3 +333,49 @@ func (w *Wallet) VerifySignature(args *VerifySignatureArgs) (*VerifySignatureRes
 func AnyoneKey() (*ec.PrivateKey, *ec.PublicKey) {
 	return ec.PrivateKeyFromBytes([]byte{1})
 }
+
+// CreateHmac generates an HMAC (Hash-based Message Authentication Code) for the provided data
+// using a symmetric key derived from the protocol, key ID, and counterparty.
+func (w *Wallet) CreateHmac(args CreateHmacArgs) (*CreateHmacResult, error) {
+	if args.Counterparty.Type == CounterpartyUninitialized {
+		args.Counterparty = Counterparty{
+			Type: CounterpartyTypeSelf,
+		}
+	}
+
+	key, err := w.keyDeriver.DeriveSymmetricKey(args.ProtocolID, args.KeyID, args.Counterparty)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive symmetric key: %w", err)
+	}
+
+	mac := hmac.New(sha256.New, key.ToBytes())
+	mac.Write(args.Data)
+	hmac := mac.Sum(nil)
+
+	return &CreateHmacResult{Hmac: hmac}, nil
+}
+
+// VerifyHmac verifies that the provided HMAC matches the expected value for the given data.
+// The verification uses the same protocol, key ID, and counterparty that were used to create the HMAC.
+func (w *Wallet) VerifyHmac(args VerifyHmacArgs) (*VerifyHmacResult, error) {
+	if args.Counterparty.Type == CounterpartyUninitialized {
+		args.Counterparty = Counterparty{
+			Type: CounterpartyTypeSelf,
+		}
+	}
+
+	key, err := w.keyDeriver.DeriveSymmetricKey(args.ProtocolID, args.KeyID, args.Counterparty)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive symmetric key: %w", err)
+	}
+
+	mac := hmac.New(sha256.New, key.ToBytes())
+	mac.Write(args.Data)
+	expectedHmac := mac.Sum(nil)
+
+	if !hmac.Equal(expectedHmac, args.Hmac) {
+		return nil, errors.New("HMAC is not valid")
```

Comment:
I would suggest to make this error public
like:
```go
var ErrHMACInvalid = errors.New("HMAC is not valid")
```

this way, adopter can distinguish between some failure (for example on key derivation)
and the situation that HMAC is really invalid, by using `errors.Is(err, wallet.ErrHMACInvalid)`

Reply:
This was changed to use `return &VerifyHmacResult{Valid: false}, nil` instead of error.